tests/test_charm.py

```diff
@@ -51,6 +51,48 @@ def test_website_relation_joined(self, open, ingress_address):
         self.assertEqual(data["private-address"], "192.168.1.17")
         self.assertEqual(data["port"], '17070')
 
+    @patch("builtins.open", new_callable=mock_open, read_data=agent_conf)
+    def test_metrics_endpoint_relation(self, open):
+        harness = Harness(JujuControllerCharm)
+        self.addCleanup(harness.cleanup)
+        harness.begin()
+
+        harness.add_network(address="192.168.1.17", endpoint="metrics-endpoint")
+
+        mock_opener = mockOpener(self)
+        harness.charm.control_socket.opener = mock_opener
+
+        mock_opener.expect(
+            url='http://localhost/metrics-users',
+            method='POST',
+            body_regex=r'{"username": "juju-metrics-r\d+", "password": ".*"}',
+            response=mockResponse(
+                headers=mockHeaders(content_type='application/json'),
+                body='{"message":"created user \\"juju-metrics-r0\\""}'
+            )
+        )
+        relation_id = harness.add_relation('metrics-endpoint', 'prometheus-k8s')
+
+        # Check metrics endpoint provider config
+        scrape_config = harness.charm.metrics_endpoint._scrape_jobs[0]
```

Comment:
This makes the test fragile, as we're testing using `_scrape_jobs`, an internal implementation detail of the prometheus_scrape charm lib. We could use `Harness.get_relation_data` to ensure the charm lib set the correct relation data. However, we don't really need to test the charm lib here (that's tested elsewhere), just that `MetricsEndpointProvider()` is called with the right data, and that `set_scrape_job_spec` is called. Probably can use a `MagicMock` for this.

Reply:
agree